Add filtering for watched objects based on label selectors and/or annotations

bellpr-work opened the pull request:
I noticed that ingress-merge was entering its processing loop every two seconds on our cluster. Some digging revealed this to be due to the ALB ingress controller using a ConfigMap for locking (which it updates every two seconds). I added CLI options to filter ConfigMaps and Ingresses based on a selector string or to ignore them based on the presence of an annotation. I also added an extremely long explicit timeout to the watches, as they will otherwise err with an "old resource version" error when the watch connection ends and it reconnects. Finally, I added some comments to the helm values file and tweaked the compile options to strip debug info (doesn't affect backtraces).

Thanks a lot for writing this, it looks perfect to bridge the gap until the ALB controller handles Ingress grouping automatically.

Commit:
Added command-line parameters to filter the monitored ConfigMap and Ingress objects based on label selectors and/or annotations. This avoids the service being woken constantly due to services like the ALB ingress controller that use a ConfigMap for locking which is updated every two seconds.
Also added a timeout parameter to the ListWatch to avoid spurious "watch of <ResourceType> ended with: too old resource version" errors whenever the watch connection ended and was recreated.

Commit:
Added new filtering switches into Helm template.
Modified Dockerfile build to omit debugging info (slight space savings)

Commit:
Added explanatory comments to Helm values template.
Added informational messages at startup reflecting any configured filters or blacklists.

Commit:
Oops, meant to process updates if *either* old or new object does not contain the blacklisted annotation. Fixed logic.

jakubkulhan commented on controller.go:
```diff
@@ -68,8 +72,17 @@ func (c *Controller) Run(ctx context.Context) (err error) {
+		cache.NewFilteredListWatchFromClient(c.client.ExtensionsV1beta1().RESTClient(), "ingresses", "",
+		func(options *metaV1.ListOptions) {
+			options.LabelSelector = c.IngressSelector
+			duration := int64(1<<31-1)
```
Use `math.MaxInt32` constant.

bellpr-work replied:
Done

jakubkulhan commented on controller.go:
```diff
@@ -68,8 +72,17 @@ func (c *Controller) Run(ctx context.Context) (err error) {
+
 	c.ingressesIndex, c.ingressesInformer = cache.NewIndexerInformer(
-		cache.NewListWatchFromClient(c.client.ExtensionsV1beta1().RESTClient(), "ingresses", "", fields.Everything()),
+		cache.NewFilteredListWatchFromClient(c.client.ExtensionsV1beta1().RESTClient(), "ingresses", "",
```
Reformat with every argument on a new line.

bellpr-work replied:
Done.

jakubkulhan commented on controller.go:
```diff
@@ -78,8 +91,18 @@ func (c *Controller) Run(ctx context.Context) (err error) {
+
 	c.configMapsIndex, c.configMapsInformer = cache.NewIndexerInformer(
-		cache.NewListWatchFromClient(c.client.CoreV1().RESTClient(), "configmaps", "", fields.Everything()),
+		cache.NewFilteredListWatchFromClient(c.client.CoreV1().RESTClient(), "configmaps", "",
```
ditto

bellpr-work replied:
Done.

jakubkulhan commented on controller.go:
```diff
@@ -78,8 +91,18 @@ func (c *Controller) Run(ctx context.Context) (err error) {
+		cache.NewFilteredListWatchFromClient(c.client.CoreV1().RESTClient(), "configmaps", "",
+		func(options *metaV1.ListOptions) {
+			options.LabelSelector = c.ConfigMapSelector
+			duration := int64(1<<31-1)
```
ditto

bellpr-work replied:
Done.

jakubkulhan commented on controller.go:
```diff
@@ -93,6 +116,22 @@ func (c *Controller) Run(ctx context.Context) (err error) {
 	}
 
+	if c.IngressSelector != "" {
+	glog.Infof("Watching Ingress objects matching the following label selector: %v", c.IngressSelector)
```
Should be indented, run `go fmt`.

bellpr-work replied:
Done.

jakubkulhan commented on cmd/ingress-merge/main.go:
```diff
@@ -72,6 +88,30 @@ func main() {
+	)
+
+	rootCmd.Flags().StringArray(
+		"ingress-blacklist",
```
I think the argument would be better named `ingress-watch-ignore`. Same goes for configmaps & related fields.

bellpr-work replied:
Done. Now it's --ingress-watch-ignore/IngressWatchIgnore/ingressWatchIgnore same with ConfigMaps.

jakubkulhan commented on controller.go:
```diff
@@ -68,8 +72,17 @@ func (c *Controller) Run(ctx context.Context) (err error) {
 	}()
 
+	if _, err = labels.Parse(c.IngressSelector) ; err != nil {
+		return fmt.Errorf("Invalid Ingress selector - %s", err)
```
Please add dependency on https://github.com/pkg/errors and use `errors.Wrap(err, "parse ingress selector failed")`

bellpr-work replied:
Done. I see there's another return fmt.error "for could not sync cache", but I wasn't sure if you wanted that wrapped as well, so I left it alone.

jakubkulhan commented on controller.go:
```diff
@@ -113,16 +152,42 @@ func (c *Controller) Run(ctx context.Context) (err error) {
 	}
 }
 
+func (c *Controller) checkBlacklist(obj interface{}) bool {
```
`checkBlacklist` -> `isIgnored`

bellpr-work replied:
Done.

jakubkulhan commented on controller.go:
```diff
@@ -113,16 +152,42 @@ func (c *Controller) Run(ctx context.Context) (err error) {
 
 func (c *Controller) OnUpdate(oldObj, newObj interface{}) {
-	c.wakeUp()
+	if !(c.checkBlacklist(oldObj) && c.checkBlacklist(newObj)) {
```
`!c.isIgnored(oldObj) || !c.isIgnored(newObj)`

bellpr-work replied:
Done.

jakubkulhan commented on helm/values.yaml:
```diff
@@ -1,5 +1,20 @@
 
+# Label selector for ConfigMap objects to be monitored for changes
+# e.g. "merge.ingress.kubernetes.io=owned"
+configmapSelector: ""
```
`configmap` -> `configMap`

bellpr-work replied:
Done.

jakubkulhan commented on helm/values.yaml:
```diff
@@ -1,5 +1,20 @@
+
+# List of annotations that will cause a ConfigMap to be ignored if present
+# e.g. "control-plane.alpha.kubernetes.io/leader"
+configmapBlacklist: []
```
`configmap` -> `configMap`

bellpr-work replied:
Done.

jakubkulhan reviewed (changes requested):
Thank you for the PR! I've added some comments. After the changes, please format code using `go fmt`:

```sh
go fmt $(go list ./... | grep -v vendor)
```

Commit:
Renamed checkBlacklist function to isIgnored
Renamed "blacklist" options and related variables to "watch-ignore"
Added github.com/pkg/errors for reporting selector errors
Replaced (1<<31-1) with math.MaxInt
Changed OnUpdate test from !(a && b) to !a || !b
Formatting changes as requested

bellpr-work commented:
I believe I've addressed all your comments. Please let me know if I missed anything. Thanks!

jakubkulhan commented:
Merged. Thank you!
